### PR TITLE
Syst evt def

### PR DESCRIPTION
**Pull request by valeriadamante**

### CODE REVIEW

File: Analysis/SystSkimmer.h [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+#include <typeindex>
+
+#include "EntryQueue.h"
+
+namespace analysis {
+typedef std::variant<int,float,double,bool,unsigned long long, long, unsigned long, unsigned int> MultiType;
+
+struct Entry {
+  bool valid{false};
+  int index;
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:19]:
```suggestion
```

File: Analysis/SystSkimmer.h [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      //std::cout << "TupleMaker::process: foreach started." << std::endl;
+      try {
+        ROOT::RDF::RNode df = df_in;
+        df.Foreach([&](const Args& ...args) {
+          std::shared_ptr<Entry> entry = std::make_shared<Entry>();
+          entry->ResizeVarValues(var_names.size());
+          //std::cout << "TupleMaker::process: running detail::putEntry->" << std::endl;
+          detail::putEntry(entry, 0,args...);
+          //std::cout << "TupleMaker::process: push entry->" << std::endl;
+          entry->valid = true;
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:94]:
```suggestion
```

File: Analysis/SystSkimmer.h [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      }
+      queue.SetAllDone();
+      //std::cout << "TupleMaker::process: foreach done." << std::endl;
+    });
+    //std::cout << "starting defining entryCentral" << std::endl;
+
+    df_out = df_out.Define("_entryCentral", [=](ULong64_t entryIndexShifted) {
+
+
+      std::shared_ptr<Entry> entryCentral = std::make_shared<Entry>();
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:112]:
```suggestion
      std::shared_ptr<Entry> entryCentral;
```

File: Analysis/SystSkimmer.h [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+      try {
+        static std::shared_ptr<Entry> entry = std::make_shared<Entry>();
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:115]:
```suggestion
        static std::shared_ptr<Entry> entry;
```

File: Analysis/SystSkimmer.h [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+typedef std::variant<int,float,double,bool,unsigned long long, long, unsigned long, unsigned int> MultiType;
+
+struct Entry {
+  bool valid{false};
+  int index;
+  std::vector<MultiType> var_values;
+
+  void ResizeVarValues(size_t size){
+    var_values.resize(size);
+  }
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:24]:
```suggestion
  explicit Entry(size_t size) : var_values(size) {}
```

File: Analysis/SystSkimmer.h [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+  ROOT::RDF::RNode process(ROOT::RDF::RNode df_in, ROOT::RDF::RNode df_out, const std::vector<std::string>& var_names)
+  {
+    thread = std::make_unique<std::thread>([=]() {
+      //std::cout << "TupleMaker::process: foreach started." << std::endl;
+      try {
+        ROOT::RDF::RNode df = df_in;
+        df.Foreach([&](const Args& ...args) {
+          std::shared_ptr<Entry> entry = std::make_shared<Entry>();
+          entry->ResizeVarValues(var_names.size());
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:90]:
```suggestion
          auto entry = std::make_shared<Entry>(var_names.size());
```

File: Analysis/SystSkimmer.h [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+    df_out = df_out.Define("_entryCentral", [=](ULong64_t entryIndexShifted) {
+
+
+      std::shared_ptr<Entry> entryCentral = std::make_shared<Entry>();
+
+      try {
+        static std::shared_ptr<Entry> entry = std::make_shared<Entry>();
+        //entry.ResizeVarValues(var_names.size());
+        while(!entry->valid || entry->GetValue<unsigned long long>(0)<entryIndexShifted){
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:117]:
```suggestion
        while(!entry || entry->GetValue<unsigned long long>(0)<entryIndexShifted){
```

File: Analysis/SystSkimmer.h [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+          entry = std::make_shared<Entry>(Entry());
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:118]:
```suggestion
          entry.reset();
```

File: Analysis/SystSkimmer.h [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+          //std::cout << "entry popped? " << queue.Pop(entry) << std::endl;
+          if (!queue.Pop(entry)) {
+            //std::cout << "entry popped " <<std::endl;
+          }
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:122]:
```suggestion
          if (!queue.Pop(entry)) {
            break;
          }
```

File: Analysis/SystSkimmer.h [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        }
+        //std::cout << entryIndexShifted << "\t"<< entry->GetValue<unsigned long long>(0)<<std::endl;
+        if(entry->valid && entry->GetValue<unsigned long long>(0)==entryIndexShifted){
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:125]:
```suggestion
        if(entry && entry->GetValue<unsigned long long>(0)==entryIndexShifted){
```

File: Analysis/SystSkimmer.h [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+          entryCentral=entry;
+        }
+        //std::cout << "sono uguali "<< entryIndexShifted << "\t"<< entryCentral.GetValue<unsigned long long>(0)<<std::endl;
+      } catch (const std::exception& e) {
+        std::cout << "Error: " << e.what() << std::endl;
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:130]:
```suggestion
        std::cout << "Error: " << e.what() << std::endl;
        throw;
```

File: Analysis/SkimEvents.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+  #print(','.join(f'"{c}"' for c in colNames))
+  #df_out = df_out.Range(nEvents)
+  #print(df_in.Describe())
+  #print(df_out.Count().GetValue())
+  colNames_v = Utilities.ListToVector(colNames)
+  tuple_maker = ROOT.analysis.TupleMaker(*col_types)(4)
+  print("tuplemaker created")
+  df_out = tuple_maker.process(ROOT.RDF.AsRNode(df_in), ROOT.RDF.AsRNode(df_out), colNames_v)
+  print("tuplemaker proceassed")
+  df_out = df_out.Define("isValid", "_entryCentral->valid")
```
Comment (kandrosov) [inline Analysis/SkimEvents.py:52]:
```suggestion
  df_out = df_out.Define("isValid", "_entryCentral.use_count() > 0")
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Analysis/SystSkimmer.h [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+      std::cout << "TupleMaker::processIn: thread started." << std::endl;
+      {
+        std::unique_lock<std::mutex> lock(mutex);
+        cond_var.wait(lock);
+      }
+      std::cout << "TupleMaker::processIn: starting foreach." << std::endl;
+      try {
+        ROOT::RDF::RNode df = df_node;
+        df.Foreach([&](const std::shared_ptr<Entry>& entry) {
+          // std::cout << "Pushing" <<std::endl;
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:83]:
```suggestion
```

File: Analysis/SystSkimmer.h [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+          if(!queue.Push(entry)) {
+            throw StopLoop();
+          }
+          // std::cout << "Pushed" <<std::endl;
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:87]:
```suggestion
```

File: Analysis/SystSkimmer.h [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+    };
+    df_out = df_out.Define("_entryCentral", [=](ULong64_t entryIndexShifted) {
+      std::shared_ptr<Entry> entryCentral;
+      try {
+        static bool notified = notify();
+        static std::shared_ptr<Entry> entry;
+        static std::set<unsigned long long> processedEntries;
+        if(processedEntries.count(entryIndexShifted))
+          throw std::runtime_error("Entry already processed");
+        // std::cout << "Poping" <<std::endl;
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:117]:
```suggestion
```

File: Analysis/SystSkimmer.h [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+          if(entry){
+            processedEntries.insert(entry->GetValue<unsigned long long>(0));
+            entry.reset();
+          }
+          if (!queue.Pop(entry)) {
+            break;
+          }
+        }
+        // std::cout << "Poped" <<std::endl;
+        //std::cout << entryIndexShifted << "\t"<< entry->GetValue<unsigned long long>(0)<<std::endl;
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:128]:
```suggestion
```

File: Analysis/SystSkimmer.h [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+        if(entry && entry->GetValue<unsigned long long>(0)==entryIndexShifted){
+          entryCentral=entry;
+          // std::cout << "Set" <<std::endl;
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:131]:
```suggestion
```

File: Analysis/SystSkimmer.h [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+        }
+        //std::cout << "sono uguali "<< entryIndexShifted << "\t"<< entryCentral->GetValue<unsigned long long>(0)<<std::endl;
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:133]:
```suggestion
```

File: Analysis/SkimEvents.py [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+  df_out_valid = df_out_valid.Define("condition_Valid", condition_Valid).Filter("condition_Valid")
+
+  snaps = []
+  print("start making screenshot")
+  snapshotOptions = ROOT.RDF.RSnapshotOptions()
+  snapshotOptions.fOverwriteIfExists=False
+  snapshotOptions.fLazy=True
+  snapshotOptions.fMode="RECREATE"
+  snapshotOptions.fCompressionAlgorithm = getattr(ROOT.ROOT, 'k' + 'LZ4')
+  snapshotOptions.fCompressionLevel = 5
```
Comment (kandrosov) [inline Analysis/SkimEvents.py:159]:
```suggestion
  snapshotOptions.fCompressionLevel = 4
```

File: Analysis/SkimEvents.py [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+  if os.path.exists(outFile_Valid):
+    os.remove(outFile_Valid)
+  if os.path.exists(outFile_nonValid):
+    os.remove(outFile_nonValid)
+  if os.path.exists(outFile_Valid_noDiff):
+    os.remove(outFile_Valid_noDiff)
+  #df_out_valid.Snapshot(f"Events", outFile_Valid, colToSave_diff_v, snapshotOptions)
+  snaps.append(df_out_valid_noDiff.Snapshot(f"Events", outFile_Valid_noDiff, colToSave_noDiff_v, snapshotOptions))
+  snaps.append(df_out_valid.Snapshot(f"Events", outFile_Valid, colToSave_diff_v, snapshotOptions))
+  snaps.append(df_unique.Snapshot(f"Events", outFile_nonValid, colNames_v, snapshotOptions))
```
Comment (kandrosov) [inline Analysis/SkimEvents.py:175]: change tree names

File: Analysis/SkimEvents.py [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+  parser.add_argument('--inFileShifted', required=True, type=str)
+  parser.add_argument('--outFile', required=True, type=str)
+  args = parser.parse_args()
+  headers_dir = os.path.dirname(os.path.abspath(__file__))
+  ROOT.gROOT.ProcessLine(f".include {os.environ['ANALYSIS_PATH']}")
+  header_path_Skimmer = os.path.join(headers_dir, "SystSkimmer.h")
+  ROOT.gInterpreter.Declare(f'#include "{header_path_Skimmer}"')
+  make_df(args.inFileCentral,args.inFileShifted,args.outFile)
+  #df_in = ROOT.RDataFrame("Events", "output/tmp_ggR_320/Events.root")
+  #df_out = ROOT.RDataFrame("Events", "output/tmp_ggR_320/Events_TotalUp.root")
```
Comment (kandrosov) [inline Analysis/SkimEvents.py:198]:
```suggestion
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: AnaProd/anaTupleProducer.py [hunk 22]
```diff
@@ -231,8 +241,8 @@ def createAnatuple(inFile, outDir, config, sample_name, anaCache, snapshotOption
     parser.add_argument('--outDir', required=True, type=str)
     parser.add_argument('--sample', required=True, type=str)
     parser.add_argument('--anaCache', required=True, type=str)
-    parser.add_argument('--compressionLevel', type=int, default=9)
-    parser.add_argument('--compressionAlgo', type=str, default="LZMA")
+    parser.add_argument('--compressionLevel', type=int, default=5)
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/anaTupleProducer.py:244]: ok

File: Analysis/SkimEvents.py [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+#import Common.Utilities as Utilities
+import os
+import shutil
+ROOT.gROOT.SetBatch(True)
+ROOT.EnableThreadSafety()
+
+ROOT.gInterpreter.Declare(
+  """
+
+namespace detail {
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Analysis/SkimEvents.py:11]: ok

File: AnaProd/anaTupleProducer.py [hunk 24]
```diff
@@ -66,6 +66,12 @@ def addAllVariables(dfw, syst_name, isData, trigger_class):
     dfw.DefineAndAppend(f"met_phi_nano", f"static_cast<float>(MET_p4_nano.phi())")
     dfw.DefineAndAppend("met_pt", "static_cast<float>(MET_p4.pt())")
     dfw.DefineAndAppend("met_phi", "static_cast<float>(MET_p4.phi())")
+    dfw.DefineAndAppend(f"nonOverlappingJets", "RemoveOverlaps(Jet_p4, Jet_bIncl,{{httCand.leg_p4[0], httCand.leg_p4[1],HbbCandidate.leg_p4[0],HbbCandidate.leg_p4[1]},}, 2, 0.5)")
+    dfw.DefineAndAppend(f"nonOverlappingJets_idx", f"Jet_idx[nonOverlappingJets]")
+    dfw.DefineAndAppend(f"nonOverlappingJets_pt", f"Jet_pt[nonOverlappingJets]")
+    dfw.DefineAndAppend(f"nonOverlappingJets_eta", f"Jet_eta[nonOverlappingJets]")
+    dfw.DefineAndAppend(f"nonOverlappingJets_phi", f"Jet_phi[nonOverlappingJets]")
+    dfw.DefineAndAppend(f"nonOverlappingJets_mass", f"Jet_mass[nonOverlappingJets]")
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply AnaProd/anaTupleProducer.py:74]: ok

File: Analysis/SkimEvents.py [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+    condition_noDiff_list.append(f"analysis::IsSame(_entryCentral->GetValue<{col_type_dict[col_types[var_idx]]}>({var_idx}),{var_name})")
+    #condition_Valid_list.append(f"! analysis::IsSame(_entryCentral->GetValue<{col_type_dict[col_types[var_idx]]}>({var_idx}),{var_name})")
+
+  condition_noDiff = ' && '.join(condition_noDiff_list)
+  df_out_valid = df_out_valid.Define("isSame", condition_noDiff)
+  df_out_valid_noDiff = df_out_valid.Filter("isSame")
+  df_out_valid_diff = df_out_valid.Filter("!isSame")
+  for var_idx,var_name in enumerate(colNames):
+    if var_name in colToNotToMakeDiff: continue
+    df_out_valid_diff=df_out_valid_diff.Define(f"diff_{var_name}", f"""analysis::Delta(_entryCentral->GetValue<{col_type_dict[col_types[var_idx]]}>({var_idx}),{var_name})""")
```
Comment (kandrosov) [inline Analysis/SkimEvents.py:56]: `{var_name}Diff` to make the naming convention "nano-friendly"

File: Analysis/SystSkimmer.h [hunk 26]
```diff
@@ -0,0 +1,197 @@
...
+template<typename T>
+T Delta(const T& shifted, const T& central)
+{
+  return detail::DeltaImpl<T>::Delta(shifted, central);
+}
+
+
+
+  template<typename T>
+  T FromDelta(T delta, T central)
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:192]: from delta needs specialisations for bool and vectors

File: Analysis/SystSkimmer.h [hunk 27]
```diff
@@ -0,0 +1,197 @@
...
+
+namespace detail {
+  template<typename T>
+  struct DeltaImpl {
+    static T Delta(const T& shifted, const T& central) {
+      return shifted - central;
+    }
+  };
+
+  template<typename T>
```
Comment (kandrosov) [inline Analysis/SystSkimmer.h:144]: also need a specialisation for bool

File: Analysis/SkimEvents.py (hunk 25 above)
Comment (valeriadamante) [reply Analysis/SkimEvents.py:56]: ok

File: Analysis/SkimEvents.py [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+    colToSave_diff.append(f"diff_{var_name}")
+
+  snaps = []
+  print("start making screenshot")
+  snapshotOptions = ROOT.RDF.RSnapshotOptions()
+  snapshotOptions.fOverwriteIfExists=False
+  snapshotOptions.fLazy=True
+  snapshotOptions.fMode="RECREATE"
+  snapshotOptions.fCompressionAlgorithm = getattr(ROOT.ROOT, 'k' + 'LZMA')
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Analysis/SkimEvents.py:65]: ok

File: Analysis/SystSkimmer.h (hunk 27 above)
Comment (valeriadamante) [reply Analysis/SystSkimmer.h:144]: ok

File: Analysis/SkimEvents.py [hunk 29]
```diff
@@ -67,9 +67,9 @@ def make_df(inputFileCentral,inputFileShifted,outFile,treeName):
   colToSave_noDiff_v = ListToVector(colToSave_noDiff)
   colToSave_diff_v = ListToVector(colToSave_diff)
   colNames_v = ListToVector(colNames)
-  outFile_Valid = f"{outFile}_Diff.root"
-  outFile_nonValid = f"{outFile}_nonValid.root"
-  outFile_Valid_noDiff = f"{outFile}_noDiff.root"
+  outFile_Valid = f"{outDir}/{treeName}_Diff.root"
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Analysis/SkimEvents.py:70]: ok

File: Analysis/SkimProducer.py [hunk 30]
```diff
@@ -0,0 +1,56 @@
...
+    sh_call(cmd, True)
+  for file_syst in os.listdir(args.workingDir):
+    syst_files_to_merge.append(f'{args.workingDir}/{file_syst}')
+  if args.test: print(f for f in syst_files_to_merge)
+  hadd_str = f'hadd -f209 -j -O {args.outputFile} '
+  hadd_str += ' '.join(f for f in syst_files_to_merge)
+  if args.test: print(hadd_str)
+  sh_call([hadd_str], True)
+  for syst_file in syst_files_to_merge:
+    os.remove(syst_file)
```
[the remark this answers is not in the document]
Comment (valeriadamante) [reply Analysis/SkimProducer.py:56]: ok

**Review by kandrosov — APPROVED** [review]

**Pull request merged**